### ISSUE TITLE
squatter segfaults when using xapian

### ISSUE THREAD

**Issue by hagedose** [cyrusimap/cyrus-imapd#1807]
I compiled 3.0rc1 on CentOS 7 with xapian support, following the revised instructions (using build.sh from cyruslibs). With searchengine set to "squat", squatter works fine. Then I added the following to imapd.conf:

```
search_engine: xapian
search_index_headers: off
search_batchsize: 8192
defaultsearchtier: temp
tempsearchpartition-default: /dev/shm/cyrus_search
sync_log: on
sync_log_channels: squatter

```
When running squatter from the command line, I get this:

```
$ squatter -u lhsu
squatter[19043]: indexing mailboxes
squatter[19043]: indexing mailbox user/lhsu...
Segmentation fault

```
When using "squatter -R" from cyrus.conf, it crashes as well:

`Jan 24 12:03:22 centos7-hg kernel: squatter[19043]: segfault at 0 ip 00007fd369cca9b5 sp 00007ffda423b610 error 4 in libcyrus_imap.so.0.0.0[7fd369c48000+a9000]
`
Here's a backtrace from gdb:

```
(gdb) bt
#0  xapian_dbw_is_indexed (dbw=0x0, cyrusid=<optimized out>) at imap/xapian_wrap.cpp:444
#1  0x00007ffff75f029a in is_indexed (rx=0x6099d0, msg=0x60b7f0) at imap/search_xapian.c:1549
#2  0x00007ffff75dfd22 in search_update_mailbox (rx=0x6099d0, mailbox=0x7ffff7f83018, flags=<optimized out>) at imap/search_engines.c:229
#3  0x0000000000403a88 in index_one (name=0x609bf0 "user.lhsu", blocking=blocking@entry=1) at imap/squatter.c:266
#4  0x00000000004033e3 in do_indexer (sa=0x7fffffffe3c0) at imap/squatter.c:326
#5  main (argc=3, argv=<optimized out>) at imap/squatter.c:968
```

**Comment by rsto** [comment]
I can't reproduce this error with a clean install (e.g. in Cassandane). Are you running squatter on an existing search partition? Or is this a fresh install? Could you please run squatter with the "-v" verbose option?

From the trace I'm tempted to blame some lines along here: https://github.com/cyrusimap/cyrus-imapd/blob/master/imap/search_xapian.c#L1463

Your search text receiver didn't get initialized with the underlying Xapian database, and ```begin_mailbox_update``` might hide this fact with a 0 return error. I'm just guessing, though, currently. Could you run squatter in gdb and breakpoint in ```begin_mailbox_update``` to validate my assumption?

**Comment by hagedose** [comment]
This is a fresh install, and it is entirely possible I haven't initialized Xapian properly. I just followed the instructions here:

https://raw.githubusercontent.com/cyrusimap/cyrus-imapd/809a6acd152e7bc61a355d563d9d72ca94d757c8/doc/README.xapian

Adding -v doesn't change anything:

```
$ squatter -v -u lhsu
squatter[19231]: indexing mailboxes
squatter[19231]: indexing mailbox user/lhsu...
Indexing mailbox user/lhsu... Segmentation fault

```
Here's my gdb session with the breakpoint:

```
$ gdb --args squatter -v -u lhsu
GNU gdb (GDB) Red Hat Enterprise Linux 7.6.1-94.el7
Copyright (C) 2013 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.  Type "show copying"
and "show warranty" for details.
This GDB was configured as "x86_64-redhat-linux-gnu".
For bug reporting instructions, please see:
<http://www.gnu.org/software/gdb/bugs/>...
Reading symbols from /usr/sbin/squatter...done.
(gdb) break begin_mailbox_update
Function "begin_mailbox_update" not defined.
Make breakpoint pending on future shared library load? (y or [n]) y
Breakpoint 1 (begin_mailbox_update) pending.
(gdb) run
Starting program: /usr/sbin/squatter -v -u lhsu
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/lib64/libthread_db.so.1".
squatter[19235]: indexing mailboxes
squatter[19235]: indexing mailbox user/lhsu...
Indexing mailbox user/lhsu...
Breakpoint 1, begin_mailbox_update (rx=0x608550, mailbox=0x7ffff7f83018, flags=0) at imap/search_xapian.c:1445
1445	imap/search_xapian.c: No such file or directory.
Missing separate debuginfos, use: debuginfo-install cyrus-sasl-lib-2.1.26-20.el7_2.x86_64 glibc-2.17-157.el7_3.1.x86_64 keyutils-libs-1.5.8-3.el7.x86_64 krb5-libs-1.14.1-27.el7_3.x86_64 libcom_err-1.42.9-9.el7.x86_64 libgcc-4.8.5-11.el7.x86_64 libselinux-2.5-6.el7.x86_64 libstdc++-4.8.5-11.el7.x86_64 nss-softokn-freebl-3.16.2.3-14.4.el7.x86_64 openssl-libs-1.0.1e-60.el7.x86_64 pcre-8.32-15.el7_2.1.x86_64 zlib-1.2.7-17.el7.x86_64
(gdb)
```

Let me know if you need anything else.

**Comment by rsto** [comment]
Hm, your build is missing debug info...

I see that you are using unix hierarchy separators (the default switched in 3.0 from off to on):

    squatter[19231]: indexing mailbox user/lhsu...

You might want to retest using turning them off with

    altnamespace : no
    unixhierarchysep : no

But toggling them didn't cause a segfault in my environment.

How does your search directory and activefiles look like? E.g. for my test environment with test user ```cassandane``` it looks like:

    $ ls /dev/shm/cyrus_search/c/user/cassandane/xapian/
    cyrus.indexed.db  cyrus.indexed.db-lock  flintlock  iamglass  position.glass  postlist.glass termlist.glass

and 

    $ cat /var/tmp/cass/1311351/conf/user/c/cassandane.xapianactive*
    temp:0

Sorry for the poking in the dark, it's tricky to debug when my squatter refuses to crash with the same issue :)

**Comment by hagedose** [comment]
How do I add debug info to the build? CFLAGS=-g?

We have always used altnamespace and unixhierarchysep, so I don't see the point in turning them off.

/dev/shm/cyrus_search is empty.

$ cat /var/lib/imap/user/C/lhsu.xapianactive
:0

**Comment by rsto** [comment]
Yes "-g".

This is what I'm typically calling ./configure with for development:

     CFLAGS="-g -fPIC -W -Wall -Wextra -Werror" \
     CC="gcc" \
     CXX="g++" \
     XAPIAN_CONFIG="$CYRUSLIBS/bin/xapian-config" \
     ./configure

/dev/shm/cyrus_search being empty is a strong indicator that ```xapian_dbw_open``` never was called or finished properly. Best guess still is that, for yet unknown reasons, ```begin_mailbox_update``` fails silently.

**Comment by hagedose** [comment]
Thanks. I rebuilt using those settings, and I can see them being used during make, but for some reason  gdb still doesn't find the source:

```
Breakpoint 1, begin_mailbox_update (rx=0x608550, mailbox=0x7ffff7f83018, flags=0) at imap/search_xapian.c:1446
1446	imap/search_xapian.c: No such file or directory.
```

Any ideas?  Ah, I noticed the path is relative! When I run gdb from the source directory, it works:

```
(gdb) run
Starting program: /usr/sbin/squatter -v -u lhsu
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/lib64/libthread_db.so.1".
squatter[26246]: indexing mailboxes
squatter[26246]: indexing mailbox user/lhsu...
Indexing mailbox user/lhsu...
Breakpoint 1, begin_mailbox_update (rx=0x6099d0, mailbox=0x7ffff7f83018, flags=0) at imap/search_xapian.c:1446
1446	    xapian_update_receiver_t *tr = (xapian_update_receiver_t *)rx;
(gdb)
```
Is there a way to force absolute paths for debug info? And what do I do now?

**Comment by rsto** [comment]
When you hit the breakpoint, please check if this function executes properly.

You can navigate to the next line (without stepping into functions) with the command ```n```. You should be able to print the numeric value of the return code variable ```r``` within that function using ```print r```

Does the control flow use one of the "goto out" jumps before hitting line 1501? If so, which one?  

Especially, do both these blocks pass?

Block 1 at [line search_xapian.c:1489](https://github.com/cyrusimap/cyrus-imapd/blob/master/imap/search_xapian.c#L1489)

        active = activefile_open(mailbox->name, mailbox->part, &tr->activefile, /*write*/1);
        if (!active || !active->count) goto out;

And block 2 at [line search_xapian.c:1493](https://github.com/cyrusimap/cyrus-imapd/blob/master/imap/search_xapian.c#L1493)

        tr->activedirs = activefile_resolve(mailbox->name, mailbox->part, active, /*dostat*/2);
        if (!tr->activedirs || !tr->activedirs->count) goto out; 

These are the lines where I can imagine a failure that's silently discarded and resulting in the dbw property of your search_text_receiver being NULL.

**Comment by hagedose** [comment]
Thanks. The first block passed, the second didn't.

```
(gdb)
1493	    tr->activedirs = activefile_resolve(mailbox->name, mailbox->part, active, /*dostat*/2);
(gdb) list
1488	     */
1489	    active = activefile_open(mailbox->name, mailbox->part, &tr->activefile, /*write*/1);
1490	    if (!active || !active->count) goto out;
1491
1492	    /* doesn't matter if the first one doesn't exist yet, we'll create it */
1493	    tr->activedirs = activefile_resolve(mailbox->name, mailbox->part, active, /*dostat*/2);
1494	    if (!tr->activedirs || !tr->activedirs->count) goto out;
1495
1496	    /* create the directory if needed */
1497	    r = check_directory(strarray_nth(tr->activedirs, 0), tr->super.verbose, /*create*/1);
(gdb) n
1494	    if (!tr->activedirs || !tr->activedirs->count) goto out;
(gdb)
1518	    free(fname);
(gdb) list
1513	    if (r) goto out;
1514
1515	    tr->super.mailbox = mailbox;
1516
1517	out:
1518	    free(fname);
1519	    strarray_free(active);
1520	    return r;
1521	}
1522
(gdb) print r
$1 = 0
```

**Comment by hagedose** [comment]
The underlying problem appears to be that in activefile_path() basedir is not set.

```
(gdb) break activefile_path
Breakpoint 2 at 0x7ffff75f2d0a: file imap/search_xapian.c, line 345.
(gdb) continue
Continuing.
Breakpoint 2, activefile_path (mboxname=0x609d20 "user.lhsu", part=0x609d60 "imap15", item=0x609e10 ":0", dostat=0) at imap/search_xapian.c:345
345	    char *basedir = NULL;
(gdb) n
346	    struct buf buf = BUF_INITIALIZER;
(gdb) n
347	    char *dest = NULL;
(gdb) n
348	    struct activeitem *ai = activeitem_parse(item);
(gdb) n
350	    xapian_basedir(ai->tier, mboxname, part, NULL, &basedir);
(gdb) n
351	    if (!basedir) goto out;
(gdb) n
370	    buf_free(&buf);
```

**Comment by rsto** [comment]
I also find your xapianactive curious, since it has an empty tier:

    $ cat /var/lib/imap/user/C/lhsu.xapianactive
    :0

It should look like

    temp:0

**Comment by hagedose** [comment]
Perhaps my imapd.conf is wrong? Does this look OK?

defaultsearchtier: temp
tempsearchpartition-default: /dev/shm/cyrus_search

**Comment by rsto** [comment]
I use the verbatim config characters in my test environment.

I am almost certain now that in your environment, this line [search_xapian:259](https://github.com/cyrusimap/cyrus-imapd/blob/master/imap/search_xapian.c#L259)

    const char *tier = config_getstring(IMAPOPT_DEFAULTSEARCHTIER);

returns an empty string tier.

Which, yeah, looks as if your installation doesn't pick up the defaultsearchtier config property. Weird.

**Comment by hagedose** [comment]
Hm, I tried to break on the line or in _activefile_init(), but the breakpoints never trigger.

**Comment by rsto** [comment]
Have you wiped /dev/shm/cyrus_search? It will only trigger once, when the file doesn't exist.

Assuming of course, that wiping data in your test environment is OK :)

**Comment by hagedose** [comment]
/dev/shm/cyrus_search is empty. I don't think anything has ever been in there ...

**Comment by rsto** [comment]
Sorry, I meant to wipe the existing activefile /var/lib/imap/user/C/lhsu.xapianactive
My bad

**Comment by hagedose** [comment]
Well, I did, but your guess appears to be wrong:

```
Breakpoint 1, _activefile_init (mboxname=0x609d20 "user.lhsu", partition=0x609d60 "imap15", activefile=0x609b80) at imap/search_xapian.c:258
258	    int r = mappedfile_writelock(activefile);
(gdb) n
259	    const char *tier = config_getstring(IMAPOPT_DEFAULTSEARCHTIER);
(gdb) n
260	    strarray_t *list = NULL;
(gdb) print tier
$2 = 0x60b640 "temp"
(gdb) continue
Continuing.

Program received signal SIGSEGV, Segmentation fault.
```

**Comment by hagedose** [comment]
But at least this has changed:

# cat /var/lib/imap/user/C/lhsu.xapianactive
temp:0

**Comment by hagedose** [comment]
OK, I'm making progress.

In xapian_rootdir(), confkey is tempsearchpartition-imap15, and of course I haven't defined that. I didn't realize I need to have one of these for each partition. Let me try to fix that.

**Comment by hagedose** [comment]
OK, that was it:

```
squatter -v -u lhsu
squatter[26403]: indexing mailboxes
squatter[26403]: indexing mailbox user/lhsu...
Indexing mailbox user/lhsu... squatter[26403]: Building directory /dev/shm/cyrus_search/C/user/lhsu/xapian
squatter[26403]: Xapian committed 199 updates in 0.082757 sec
squatter[26403]: write_indexed db=/dev/shm/cyrus_search/C/user/lhsu/xapian/cyrus.indexed.db mailbox=user.lhsu uidvalidity=1456139918 uids=1:199
squatter[26403]: write_indexed db=/dev/shm/cyrus_search/C/user/lhsu/xapian/cyrus.indexed.db mailbox=user.lhsu uidvalidity=1456139918 uids=1:199
squatter[26403]: done indexing mailboxes
```

Still that configuration error should be handled better. Thanks for your help!

**Comment by rsto** [comment]
Yeah, you stumbled along a code path where return codes weren't nicely handled. Thanks for catching this.

**Comment by rsto** [comment]
We report this error now in the syslogs https://github.com/cyrusimap/cyrus-imapd/commit/ca4b76ef37accff7b1f5c3e440067a92920d3575

**Comment by postilion** [comment]
Robert,
Could you please send me a note of guidance for documentation on this issue?  I'd like to get that into the docs sooner rather than later.  Right now we have very little for Xapian docs, other than the developer notes (docsrc/imap/developer/install-xapian.rst).

If I understand this correctly, we need to highlight the interdependence of these configuration settings:

    defaultsearchtier: temp
    tempsearchpartition-default: /dev/shm/cyrus_search

It sounds like `hagedose' was missing a set of parameters like this for each partition, such that if he had this in his configuration:

    partition-1: /var/mailstores/1
    partition-2: /var/mailstores/2
    partition-3: /var/mailstores/3
    partition-4: /var/mailstores/4
    partition-5: /var/mailstores/5
    partition-6: /var/mailstores/6
    partition-7: /var/mailstores/7
    partition-8: /var/mailstores/8
    partition-9: /var/mailstores/9
    partition-10: /var/mailstores/10

He would also need this:

    tempsearchpartition-1: /dev/shm/yadda
    tempsearchpartition-2: /dev/shm/yadda
    tempsearchpartition-3: /dev/shm/yadda
    tempsearchpartition-4: /dev/shm/yadda
    tempsearchpartition-5: /dev/shm/yadda
    tempsearchpartition-6: /dev/shm/yadda
    tempsearchpartition-7: /dev/shm/yadda
    tempsearchpartition-8: /dev/shm/yadda
    tempsearchpartition-9: /dev/shm/yadda
    tempsearchpartition-10: /dev/shm/yadda

Is that correct?  Then what should "yadda" be?  Is it the same for each, or must it be different for each, or what?  Since <tier>searchpartition-<name> is not documented in /lib/imapoptions, I've got nothing to go on here.

Cheers,
    -nic

**Comment by postilion** [comment]
To expand on this a bit further, the entry in imapoptions for ``defaultsearchtier`` reads:

> Name of the default tier that messages will be indexed to 

Is that correct, or does this mean the tier to which messages in the partition "default" will be indexed?

It's unfortunate that in so many places the name "default" is used as a placekeeping example -- such as defaultpartition & partition-default -- when it doesn't mean "default" as one normally would mean it, but rather is just a placekeeper.  That makes it really unclear, then, when it does actually mean "default," like most people would mean it.

In other words, if I have the list of partitions, as shown above, would I also need this:
`1searchtier: yadda1`
`2searchtier: yadda2`
... etc.?

Cheers,
-nic

**Comment by rsto** [comment]
Nic, you beat me to it. Just wanted to raise an issue for docs on this. I'll just reopen this one instead.

Unfortunately, I don't have a multi-partitioned mail setup and haven't fiddled with xapian activefiles up until this issue. That being said I don't see an issue with using the same underlying Xapian database for multiple partitions of the same tier.  In your example the ```yadda``` could all be the same for all partitions. CCing @brong to have a final say in this.

The defaultsearch tier is the one tier which will be used to index new messages to. The 4-layered FastMail tier setup in this [https://blog.fastmail.com/2014/12/01/email-search-system/](blog post) goes into more details.

I agree: the docs on Cyrus search using Xapian for 3.0 need more clarification. I'll be happy to help, but as much as I worked on the core Xapian layers I don't have much experience yet with its operational setup. Thanks for pushing this.

**Comment by elliefm** [comment]
> Then what should "yadda" be? Is it the same for each, or must it be different for each, or what? Since searchpartition- is not documented in /lib/imapoptions, I've got nothing to go on here.

n.b I'm speculating a bit here so I might be wrong

```
tempsearchpartition-1: /dev/shm/yadda
tempsearchpartition-2: /dev/shm/yadda
tempsearchpartition-3: /dev/shm/yadda
...
```

I think these may need to at least be different directories, though it's probably okay for those to be under the same mountpoint.  It *might* be okay for them to all be the same directory, but I don't know that.

So like:

```
tempsearchpartition-1: /dev/shm/cyrus_search/1
tempsearchpartition-2: /dev/shm/cyrus_search/2
tempsearchpartition-3: /dev/shm/cyrus_search/3
...
```

Interestingly, the blog post @rsto links to (2014) talks only about `[tier]searchpartition-default`, seeming to imply that searchpartition does not correspond to real partitions (unlike metapartition/archivepartition/etc), or maybe it's a cut-down example, or maybe config has changed since that post? 

I dunno, it's not clear to me at the moment whether there is supposed to be _one search partition per tier_, or whether there's supposed to be _one search partition per partition per tier_.  Looking at code, it looks like the latter.  But looking at the blog example, it looks like maybe former was intended?

**Comment by hagedose** [comment]
The main problem regarding the documentation is that the word "default" is used many times with slightly different meanings. Using trial and error I have determined that it is supposed to be one search partition per partition per tier.

Here's an annotated version of what I have right now:

```
partition-default: /var/spool/imap # A partition with the _name_ default (this should be discouraged)
partition-imap2: /var/spool/imap2 # A partition with the name imap2
...

defaultsearchtier: temp # the search tier used by default. Not clear if Fastmail's multi-tier approach is actually useful for smaller installations

tempsearchpartition-default: /dev/shm/cyrus_search # The location of the index for the partition named default in the tier named temp
tempsearchpartition-imap2: /var/lib/imap/cyrus_search2 # The location of the index for the partition named imap2 in the tier named temp
...

```
I would really welcome a more detailed discussion of using multiple tiers. The manpage for squatter doesn't yet mention the flags -z and -t used in the blog examples.

**Comment by postilion** [comment]
In reading over the source, I suspect there's a problem with how IMAPOPT_DEFAULTPARTITION is being used by search_xapian.c.  IMAPOPT_DEFAULTPARTITION derives from the imapd.conf directive 'defaultpartition:' The description of that directive states (emphasis mine):

> The partition name used by default for new mailboxes.  **If not specified,** the partition with the most free space will be used for new mailboxes.

So 'defaultpartition:' by implication may be undefined, in which case it is NULL.  Now I'm no C programmer, but it doesn't look to me like search_xapian has any special handling for that eventuality.

`    if (!partition)`
`        partition = config_getstring(IMAPOPT_DEFAULTPARTITION);`
`    confkey = strconcat(tier, "searchpartition-", partition, NULL);`
`    root = config_getoverflowstring(confkey, NULL);`

Am I right about this?  If so, do we need to change the description of 'defaultpartition:' to require a value be present?

**Comment by rsto** [comment]
Yes, this looks broken. There's some validation of IMAPOPT_DEFAULTPARTITION in global.c but also there the NULL-case isn't handled.

What happens in the code snippet

    confkey = strconcat(tier, "searchpartition-", partition, NULL);

is that for NULL-partitions this will return the confkey "searchpartition-" and that's not going to exist.

It's on my TODO for tomorrow to setup xapian from scratch on my local machine and update the docs accordingly. At the moment there are too many hidden assumptions in the codepath.

**Comment by postilion** [comment]
@rsto, thanks!  We need entries in /lib/imapoptions for searchpartition-name, for example, and probably greater clarification on defaultsearchtier there, as well.

Once that stuff is in place, we'll get handles to reference into imapd.conf(5) for use elsewhere in the doc tree, and I can write up some pretty verbiage for the various documents elsewhere.  Right now I'm working on locations.rst and its subordinate docs, and we should have explanations about Xapian stuff in there.

Cheers,
-nic

**Comment by rsto** [comment]
@postilion I've just pushed https://github.com/cyrusimap/cyrus-imapd/commit/3157bbc7c1c86b2857d8a37dcb820127648e051f to catch xapian configuration errors and be verbose about what's missing. I've also updated ```lib/imapoptions```.

I hope that this is enough for users that are OK with using just one search tier. For multi-layered search tiers I think the squatter man pages needs an update as well. I've created a follow-issue here https://github.com/cyrusimap/cyrus-imapd/issues/1826

**Comment by hagedose** [comment]
We use an empty default partition, and I don't believe it is causing problems. Here's why: xapian only comes into play when a mailbox already exists, right?. So there is always a partition associated with each mailbox. The only time the default partition is being used is when a new mailbox is created. Or am I missing something?
Looking at the source I would suggest that xapian_rootdir() can't ever be called without a value for partition so that the code path following 
`if (!partition)`
should never be executed. Of course that condition should be checked, but I don't believe it makes any sense the way it currently stands.
Please correct me if I'm wrong ...

**Comment by rsto** [comment]
Right, that looks as if it was too restrictive! I've just updated xapian search to not enforce defaultpartition to be defined, but pimped up the runtime check to safeguard future hiccups -> https://github.com/cyrusimap/cyrus-imapd/commit/bc86c894640aba6ce8bb2ef003e92f66cebb3f8b

**Comment by rsto** [comment]
With the segfault closed and proper Xapian config checks I'd say we can close this issue.
